### PR TITLE
[WIP] Support slice as `col_sel` for `loc` indexer

### LINKED ISSUE #1176: Support slice as `col_sel` for `loc` indexer.
We should support slice as `col_sel` for `loc` indexer.

```py
>>> pdf
   x     y     z
   a  b  c  d  e
0  1  2  3  4  5
>>> pdf.loc[:, 'y':]
   y     z
   c  d  e
0  3  4  5
>>> pdf.loc[:, :'y']
   x     y
   a  b  c  d
0  1  2  3  4
>>> pdf.loc[:, ('x','b'):]
   x  y     z
   b  c  d  e
0  2  3  4  5
>>> pdf.loc[:, :('y','c')]
   x     y
   a  b  c
0  1  2  3
>>> pdf.loc[:, ('x','b'):('y','c')]
   x  y
   b  c
0  2  3
>>> pdf.loc[:, 'x':('y','c')]
   x     y
   a  b  c
0  1  2  3
>>> pdf.loc[:, ('x','b'):'y']
   x  y
   b  c  d
0  2  3  4
```

### PR DESCRIPTION
**Pull request by itholic**
Resolves #1176 

```python
>>> kdf
   x     y     z
   a  b  c  d  e
0  1  2  3  4  5

>>> kdf.loc[:, 'y':]
   y     z
   c  d  e
0  3  4  5

>>> kdf.loc[:, :'y']
   x     y
   a  b  c  d
0  1  2  3  4

>>> kdf.loc[:, ('x', 'b'):]
   x  y     z
   b  c  d  e
0  2  3  4  5

>>> kdf.loc[:, :('y', 'c')]
   x     y
   a  b  c
0  1  2  3

>>> kdf.loc[:, ('x', 'b'):('y', 'c')]
   x  y
   b  c
0  2  3

>>> kdf.loc[:, 'x':('y', 'c')]
   x     y
   a  b  c
0  1  2  3

>>> kdf.loc[:, ('x', 'b'):'y']
   x  y
   b  c  d
0  2  3  4
```

### CODE REVIEW

**Comment by itholic** [comment]
While i'm here i'm going to consider below things and add more codes for handling missing exception cases with proper way.

### the things we need to solve.

pandas doesn't support slice as `col_sel` for `MultiIndex` when the first index is not monotonically increasing.

for example

```python
>>> pdf
   z  y  x
   c  b  a
0  1  2  3

>>> pdf.loc[:, 'z':]
Traceback (most recent call last):
...
pandas.errors.UnsortedIndexError: 'Key length (1) was greater than MultiIndex lexsort depth (0)'
```

so i think we need #1183 for checking if the order of column is monotonic_increasing.
(of course, there are other ways to work around, but I think maybe it's better to use the API.)

File: databricks/koalas/indexing.py [hunk 1]
```diff
@@ -555,8 +589,8 @@ def _select_cols(self, cols_sel):
             if cols_sel == slice(None):
                 cols_sel = None
             else:
-                raise LocIndexer._raiseNotImplemented(
-                    "Can only select columns either by name or reference or all")
+                return self._get_from_multiindex_column(cols_sel)
```
Comment (ueshin) [inline databricks/koalas/indexing.py]: Maybe we should just add another method to handle this case?

File: databricks/koalas/indexing.py [hunk 2]
```diff
@@ -520,7 +520,6 @@ def _get_from_multiindex_column(self, key, indexes=None):
         :param key: the multi-index column keys represented by tuple
         :return: DataFrame or Series
         """
-        assert isinstance(key, tuple)
```
Comment (ueshin) [inline databricks/koalas/indexing.py]: Shall we revert this too?

File: databricks/koalas/indexing.py [hunk 3]
```diff
@@ -530,7 +529,8 @@ def _get_from_multiindex_column(self, key, indexes=None):
 
         if all(len(idx) > 0 and idx[0] == '' for _, idx in indexes):
             # If the head is '', drill down recursively.
-            indexes = [(col, tuple([str(key), *idx[1:]])) for i, (col, idx) in enumerate(indexes)]
```
Comment (ueshin) [inline databricks/koalas/indexing.py]: ditto.

**Comment by ueshin** [comment]
@itholic I'll take this over.

**Comment by itholic** [comment]
Thanks, @ueshin . I'll take a look at that PR soo.

**Comment by itholic** [comment]
Closing. See the #1351 which continue this work.

**Pull request closed without merging**